File: algorithms/06_javascript/sorting/insertion_sort.js
function insertionSort(arr) {
  const a = arr.slice();
  for (let i = 1; i < a.length; i++) {
    const key = a[i];
    let j = i - 1;
    while (j >= 0 && a[j] > key) { a[j + 1] = a[j]; j--; }
    a[j + 1] = key;
  }
  return a;
}

module.exports = { insertionSort };

if (require.main === module) {
  const sorted = insertionSort([5, 2, 4, 6, 1, 3]);
  if (sorted.join(",") !== "1,2,3,4,5,6") throw new Error("insertion sort failed");
  console.log("[JavaScript Insertion Sort] Sorted array verified");
}

File: algorithms/22_haskell/sorting/InsertionSort.hs
module InsertionSort (insertionSort) where

insertionSort :: Ord a => [a] -> [a]
insertionSort = foldl ins []
  where
    ins [] x = [x]
    ins ys@(y:ys') x
      | x <= y = x : ys
      | otherwise = y : ins ys' x

main :: IO ()
main = do
  let data' = [33, 7, 91, 12, 5, 5, 78, 2, 44, 19] :: [Int]
      sorted = insertionSort data'
  if sorted == [2, 5, 5, 7, 12, 19, 33, 44, 78, 91]
    then putStrLn ("[Haskell InsertionSort] Insertion sort verified: " ++ show sorted)
    else error "InsertionSort verification failed"
